File: code/java/ch16/src/main/java/com/github/zxh0/luago/compiler/lexer/Lexer.java
package com.github.zxh0.luago.compiler.lexer;

import java.util.regex.Pattern;

import static com.github.zxh0.luago.compiler.lexer.TokenKind.*;

public class Lexer {

    //private static final Pattern reSpaces = Pattern.compile("^\\s+");
    private static final Pattern reNewLine = Pattern.compile("\r\n|\n\r|\n|\r");
    private static final Pattern reIdentifier = Pattern.compile("^[_\\d\\w]+");
    private static final Pattern reNumber = Pattern.compile("^0[xX][0-9a-fA-F]*(\\.[0-9a-fA-F]*)?([pP][+\\-]?[0-9]+)?|^[0-9]*(\\.[0-9]*)?([eE][+\\-]?[0-9]+)?");
    private static final Pattern reShortStr = Pattern.compile("(?s)(^'(\\\\\\\\|\\\\'|\\\\\\n|\\\\z\\s*|[^'\\n])*')|(^\"(\\\\\\\\|\\\\\"|\\\\\\n|\\\\z\\s*|[^\"\\n])*\")");
    private static final Pattern reOpeningLongBracket = Pattern.compile("^\\[=*\\[");

    private CharSeq chunk;
    private String chunkName;
    private int line;
    // to support lookahead
	private Token cachedNextToken;
	private int lineBackup;

    public Lexer(String chunk, String chunkName) {
        this.chunk = new CharSeq(chunk);
        this.chunkName = chunkName;
        this.line = 1;
    }

    public int line() {
    	return cachedNextToken != null ? lineBackup : line;
    }

    <T> T error(String fmt, Object... args) {
        String msg = String.format(fmt, args);
        msg = String.format("%s:%d: %s", chunkName, line(), msg);
        throw new RuntimeException(msg);
    }

    public TokenKind LookAhead() {
        if (cachedNextToken == null) {
            lineBackup = line;
            cachedNextToken = nextToken();
        }
        return cachedNextToken.getKind();
    }

    public Token nextIdentifier() {
        return nextTokenOfKind(TOKEN_IDENTIFIER);
    }

    public Token nextTokenOfKind(TokenKind kind) {
        Token token = nextToken();
        if (token.getKind() != kind) {
            error("syntax error near '%s'", token.getValue());
        }
        return token;
    }

    public Token nextToken() {
        if (cachedNextToken != null) {
            Token token = cachedNextToken;
            cachedNextToken = null;
            return token;
        }

        skipWhiteSpaces();
        if (chunk.length() <= 0) {
            return new Token(line, TOKEN_EOF, "EOF");
        }

        switch (chunk.charAt(0)) {
            case ';': chunk.next(1); return new Token(line, TOKEN_SEP_SEMI,   ";");
            case ',': chunk.next(1); return new Token(line, TOKEN_SEP_COMMA,  ",");
            case '(': chunk.next(1); return new Token(line, TOKEN_SEP_LPAREN, "(");
            case ')': chunk.next(1); return new Token(line, TOKEN_SEP_RPAREN, ")");
            case ']': chunk.next(1); return new Token(line, TOKEN_SEP_RBRACK, "]");
            case '{': chunk.next(1); return new Token(line, TOKEN_SEP_LCURLY, "{");
            case '}': chunk.next(1); return new Token(line, TOKEN_SEP_RCURLY, "}");
            case '+': chunk.next(1); return new Token(line, TOKEN_OP_ADD,     "+");
            case '-': chunk.next(1); return new Token(line, TOKEN_OP_MINUS,   "-");
            case '*': chunk.next(1); return new Token(line, TOKEN_OP_MUL,     "*");
            case '^': chunk.next(1); return new Token(line, TOKEN_OP_POW,     "^");
            case '%': chunk.next(1); return new Token(line, TOKEN_OP_MOD,     "%");
            case '&': chunk.next(1); return new Token(line, TOKEN_OP_BAND,    "&");
            case '|': chunk.next(1); return new Token(line, TOKEN_OP_BOR,     "|");
            case '#': chunk.next(1); return new Token(line, TOKEN_OP_LEN,     "#");
            case ':':
                if (chunk.startsWith("::")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_SEP_LABEL, "::");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_SEP_COLON, ":");
                }
            case '/':
                if (chunk.startsWith("//")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_IDIV, "//");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_OP_DIV, "/");
                }
            case '~':
                if (chunk.startsWith("~=")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_NE, "~=");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_OP_WAVE, "~");
                }
            case '=':
                if (chunk.startsWith("==")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_EQ, "==");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_OP_ASSIGN, "=");
                }
            case '<':
                if (chunk.startsWith("<<")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_SHL, "<<");
                } else if (chunk.startsWith("<=")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_LE, "<=");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_OP_LT, "<");
                }
            case '>':
                if (chunk.startsWith(">>")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_SHR, ">>");
                } else if (chunk.startsWith(">=")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_GE, ">=");
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_OP_GT, ">");
                }
            case '.':
                if (chunk.startsWith("...")) {
                    chunk.next(3);
                    return new Token(line, TOKEN_VARARG, "...");
                } else if (chunk.startsWith("..")) {
                    chunk.next(2);
                    return new Token(line, TOKEN_OP_CONCAT, "..");
                } else if (chunk.length() == 1 || !CharUtil.isDigit(chunk.charAt(1))) {
                    chunk.next(1);
                    return new Token(line, TOKEN_SEP_DOT, ".");
                }
            case '[':
                if (chunk.startsWith("[[") || chunk.startsWith("[=")) {
                    return new Token(line, TOKEN_STRING, scanLongString());
                } else {
                    chunk.next(1);
                    return new Token(line, TOKEN_SEP_LBRACK, "[");
                }
            case '\'':
            case '"':
                return new Token(line, TOKEN_STRING, scanShortString());
        }

        char c = chunk.charAt(0);
        if (c == '.' || CharUtil.isDigit(c)) {
            return new Token(line, TOKEN_NUMBER, scanNumber());
        }
        if (c == '_' || CharUtil.isLetter(c)) {
            String id = scanIdentifier();
            return Token.keywords.containsKey(id)
                    ? new Token(line, Token.keywords.get(id), id)
                    : new Token(line, TOKEN_IDENTIFIER, id);
        }

        return error("unexpected symbol near %c", c);
    }

    private void skipWhiteSpaces() {
        while (chunk.length() > 0) {
            if (chunk.startsWith("--")) {
                skipComment();
            } else if (chunk.startsWith("\r\n") || chunk.startsWith("\n\r")) {
                chunk.next(2);
                line += 1;
            } else if (CharUtil.isNewLine(chunk.charAt(0))) {
                chunk.next(1);
                line += 1;
            } else if(CharUtil.isWhiteSpace(chunk.charAt(0))) {
                chunk.next(1);
            } else {
                break;
            }
        }
    }

    private void skipComment() {
        chunk.next(2); // skip --

        // long comment ?
        if (chunk.startsWith("[")) {
            if (chunk.find(reOpeningLongBracket) != null) {
                scanLongString();
                return;
            }
        }

        // short comment
        while(chunk.length() > 0 && !CharUtil.isNewLine(chunk.charAt(0))) {
            chunk.next(1);
        }
    }

    private String scanIdentifier() {
        return scan(reIdentifier);
    }

    private String scanNumber() {
        return scan(reNumber);
    }

    private String scan(Pattern pattern) {
        String token = chunk.find(pattern);
        if (token == null) {
            throw new RuntimeException("unreachable!");
        }
        chunk.next(token.length());
        return token;
    }

    private String scanLongString() {
        String openingLongBracket = chunk.find(reOpeningLongBracket);
        if (openingLongBracket == null) {
            return error("invalid long string delimiter near '%s'",
                    chunk.substring(0, 2));
        }

        String closingLongBracket = openingLongBracket.replace("[", "]");
        int closingLongBracketIdx = chunk.indexOf(closingLongBracket);
        if (closingLongBracketIdx < 0) {
            return error("unfinished long string or comment");
        }

        String str = chunk.substring(openingLongBracket.length(), closingLongBracketIdx);
        chunk.next(closingLongBracketIdx + closingLongBracket.length());

        str = reNewLine.matcher(str).replaceAll("\n");
        line += str.chars().filter(c -> c == '\n').count();
        if (str.startsWith("\n")) {
            str = str.substring(1);
        }

        return str;
    }

    private String scanShortString() {
        String str = chunk.find(reShortStr);
        if (str != null) {
            chunk.next(str.length());
            str = str.substring(1, str.length() - 1);
            if (str.indexOf('\\') >= 0) {
                line += reNewLine.split(str).length - 1;
                str = new Escaper(str, this).escape();
            }
            return str;
        }
        return error("unfinished string");
    }

}

File: code/python/src/compiler/lexer.py
from compiler.lua_token import *
import re


class Lexer:
    re_new_line = re.compile(r"\r\n|\n\r|\n|\r")
    re_identifier = re.compile(r"^[_\d\w]+")
    re_opening_long_bracket = re.compile(r"^\[=*\[")
    re_short_string = re.compile(r"(?s)(^'(\\\\|\\'|\\\n|\\z\s*|[^'\n])*')|(^\"(\\\\|\\\"|\\\n|\\z\s*|[^\"\n])*\")")
    re_number = re.compile(r"^0[xX][0-9a-fA-F]*(\.[0-9a-fA-F]*)?([pP][+\-]?[0-9]+)?|"
                           r"^[0-9]*(\.[0-9]*)?([eE][+\-]?[0-9]+)?")

    re_dec_escape_seq = re.compile(r"^\\[0-9]{1,3}")
    re_hex_escape_seq = re.compile(r"^\\x[0-9a-fA-F]{2}")
    re_unicode_escape_seq = re.compile(r"^\\u{[0-9a-fA-F]+}")

    def __init__(self, chunk, chunk_name):
        self.chunk = chunk
        self.chunk_name = chunk_name
        self.line = 1
        self.next_token = None
        self.next_token_kind = None
        self.next_token_line = 0

    def get_line(self):
        return self.line

    def get_next_token_of_kind(self, k):
        line, kind, token = self.get_next_token()
        if k != kind:
            err = 'unexpected symbol near "{0}"'.format(token)
            self.error(err)
        return line, token

    def get_next_identifier(self):
        return self.get_next_token_of_kind(TokenKind.IDENTIFIER)

    def look_ahead(self):
        if self.next_token_line > 0:
            return self.next_token_kind
        current_line = self.line
        line, kind, token = self.get_next_token()
        self.line = current_line
        self.next_token_kind = kind
        self.next_token_line = line
        self.next_token = token
        return kind

    def get_next_token(self):
        if self.next_token_line > 0:
            line, kind, token = self.next_token_line, self.next_token_kind, self.next_token
            self.line = self.next_token_line
            self.next_token_line = 0
            return line, kind, token

        self.skip_space()
        if len(self.chunk) == 0:
            return self.line, TokenKind.EOF, 'EOF'

        c = self.chunk[0]
        if c in single_tokens:
            self.next(1)
            return self.line, single_tokens[c], c
        if c == ':':
            if self.test("::"):
                self.next(2)
                return self.line, TokenKind.SEP_LABEL, '::'
            else:
                self.next(1)
                return self.line, TokenKind.SEP_COLON, c
        if c == '/':
            if self.test('//'):
                self.next(2)
                return self.line, TokenKind.OP_IDIV, '//'
            else:
                self.next(1)
                return self.line, TokenKind.OP_DIV, c
        if c == '~':
            if self.test('~='):
                self.next(2)
                return self.line, TokenKind.OP_NE, '~='
            else:
                self.next(1)
                return self.line, TokenKind.OP_WAVE, c
        if c == '=':
            if self.test('=='):
                self.next(2)
                return self.line, TokenKind.OP_EQ, '=='
            else:
                self.next(1)
                return self.line, TokenKind.OP_ASSIGN, c
        if c == '<':
            if self.test('<<'):
                self.next(2)
                return self.line, TokenKind.OP_SHL, '<<'
            elif self.test('<='):
                self.next(2)
                return self.line, TokenKind.OP_LE, '<='
            else:
                self.next(1)
                return self.line, TokenKind.OP_LT, c
        if c == '>':
            if self.test('>>'):
                self.next(2)
                return self.line, TokenKind.OP_SHR, '>>'
            elif self.test('>='):
                self.next(2)
                return self.line, TokenKind.OP_GE, '>='
            else:
                self.next(1)
                return self.line, TokenKind.OP_GT, c
        if c == '.':
            if self.test('...'):
                self.next(3)
                return self.line, TokenKind.VARARG, '...'
            elif self.test('..'):
                self.next(2)
                return self.line, TokenKind.OP_CONCAT, '..'
            elif len(self.chunk) == 1 or not self.chunk[1].isdigit():
                self.next(1)
                return self.line, TokenKind.SEP_DOT, c
        if c == '[':
            if self.test('[[') or self.test('[='):
                return self.line, TokenKind.STRING, self.scan_long_string()
            else:
                self.next(1)
                return self.line, TokenKind.SEP_LBRACK, '['
        if c in ('\'', '"'):
            return self.line, TokenKind.STRING, self.scan_short_string()

        if c == '.' or c.isdigit():
            token = self.scan_number()
            return self.line, TokenKind.NUMBER, token
        if c == '_' or c.isalpha():
            token = self.scan_identifier()
            if token in keywords:
                return self.line, keywords[token], token
            else:
                return self.line, TokenKind.IDENTIFIER, token

        err = 'unexpected symbol near "{0}"'.format(c)
        self.error(err)

    def next(self, n):
        self.chunk = self.chunk[n:]

    def test(self, s):
        return self.chunk.startswith(s)

    def error(self, f, *args):
        err = f.format(args)
        err = '{0}:{1}: {2}'.format(self.chunk_name, self.line, err)
        raise Exception(err)

    def skip_space(self):
        while len(self.chunk) > 0:
            if self.test('--'):
                self.skip_comment()
            elif self.test('\r\n') or self.test('\n\r'):
                self.next(2)
                self.line += 1
            elif Lexer.is_new_line(self.chunk[0]):
                self.next(1)
                self.line += 1
            elif Lexer.is_white_space(self.chunk[0]):
                self.next(1)
            else:
                break

    @staticmethod
    def is_white_space(c):
        return c in ('\t', '\n', '\v', '\f', '\r', ' ')

    @staticmethod
    def is_new_line(c):
        return c in ('\r', '\n')

    def skip_comment(self):
        self.next(2)

        if self.test('['):
            if re.match(Lexer.re_opening_long_bracket, self.chunk):
                self.scan_long_string()
                return

        while len(self.chunk) > 0 and not self.is_new_line(self.chunk[0]):
            self.next(1)

    def scan_identifier(self):
        return self.scan(Lexer.re_identifier)

    def scan_short_string(self):
        m = re.match(Lexer.re_short_string, self.chunk)
        if m:
            s = m.group()
            self.next(len(s))
            s = s[1: len(s)-1]
            if s.find('\\') >= 0:
                result = Lexer.re_new_line.findall(s)
                self.line += len(result)
                s = self.escape(s)
            return s
        self.error('unfinished string')
        return ''

    def scan_long_string(self):
        m_open = re.match(Lexer.re_opening_long_bracket, self.chunk)
        if m_open is None:
            self.error('invalid long string delimiter near "%s"'.format(self.chunk[0:2]))

        str_open = m_open.group()
        str_close = str_open.replace('[', ']')
        close_idx = self.chunk.find(str_close)
        if close_idx < 0:
            self.error('unfinished long string or comment')

        s = self.chunk[len(str_open): close_idx]
        self.next(close_idx + len(str_close))

        s = re.sub(Lexer.re_new_line, s, '\n')
        self.line += s.count('\n')
        if len(s) > 0 and s[0] == '\n':
            s = s[1:]

        return s

    def scan_number(self):
        return self.scan(Lexer.re_number)

    def scan(self, pattern):
        m = re.match(pattern, self.chunk)
        if m:
            token = m.group()
            self.next(len(token))
            return token
        raise Exception('unreachable')

    def escape(self, s):
        ret = ''
        while len(s) > 0:
            if s[0] != '\\':
                ret += s[0]
                s = s[1:]
                continue

            if len(s) == 1:
                self.error('unfinished string')

            if s[1] == 'a':
                ret += '\a'
                s = s[2:]
                continue
            elif s[1] == 'b':
                ret += '\b'
                s = s[2:]
                continue
            elif s[1] == 'f':
                ret += '\f'
                s = s[2:]
                continue
            elif s[1] == 'n' or s[1] == '\n':
                ret += '\n'
                s = s[2:]
                continue
            elif s[1] == 'r':
                ret += '\r'
                s = s[2:]
                continue
            elif s[1] == 't':
                ret += '\t'
                s = s[2:]
                continue
            elif s[1] == 'v':
                ret += '\v'
                s = s[2:]
                continue
            elif s[1] == '"':
                ret += '"'
                s = s[2:]
                continue
            elif s[1] == '\'':
                ret += '\''
                s = s[2:]
                continue
            elif s[1] == '\\':
                ret += '\\'
                s = s[2:]
                continue
            elif s[1] in '0123456789':
                m = re.match(Lexer.re_dec_escape_seq, s)
                if m:
                    str_dec = m.group()[1:]
                    d = int(str_dec)
                    if d <= 0xff:
                        ret += str(chr(d))
                        s = s[len(m.group()):]
                        continue
                    self.error('decimal escape too large near "%s"'.format(m.group()))
            elif s[1] == 'x':
                m = re.match(Lexer.re_hex_escape_seq, s)
                if m:
                    str_hex = '0' + m.group()[1:]
                    d = int(str_hex, 16)
                    ret += str(chr(d))
                    s = s[len(m.group()):]
                    continue
            elif s[1] == 'u':
                m = re.match(Lexer.re_unicode_escape_seq, s)
                if m:
                    str_unicode = m.group()[3: len(m.group())-1]
                    d = int(str_unicode, 16)
                    if d <= 0x10ffff:
                        ret += str(chr(d))
                        s = s[len(m.group()):]
                        continue
                    self.error('UTF-8 value too large near "%s"'.format(str_unicode))
            elif s[1] == 'z':
                s = s[2:]
                while len(s) > 0 and Lexer.is_white_space(s[0]):
                    s = s[1:]
                continue
        return ret
